traits/editor_factories.py

```diff
@@ -0,0 +1,207 @@
+#  Copyright (c) 2005-19, Enthought, Inc.
+#  All rights reserved.
+#
+#  This software is provided without warranty under the terms of the BSD
+#  license included in enthought/LICENSE.txt and may be redistributed only
+#  under the conditions described in the aforementioned license.  The license
+#  is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+#  Thanks for using Enthought open source!
+
+from functools import partial
+
+# -------------------------------------------------------------------------------
+#  Editor factory functions:
+# -------------------------------------------------------------------------------
+
+PasswordEditors = {}
+MultilineTextEditors = {}
+BytesEditors = {}
+SourceCodeEditor = None
+HTMLTextEditor = None
+PythonShellEditor = None
+DateEditor = None
+TimeEditor = None
+
+
+def password_editor(auto_set=True, enter_set=False):
+    """ Factory function that returns an editor for passwords.
+    """
+    if (auto_set, enter_set) not in PasswordEditors:
+        from traitsui.api import TextEditor
+
+        PasswordEditors[auto_set, enter_set] = TextEditor(
+            password=True, auto_set=auto_set, enter_set=enter_set
+        )
+
+    return PasswordEditors[auto_set, enter_set]
+
+
+def multi_line_text_editor(auto_set=True, enter_set=False):
+    """ Factory function that returns a text editor for multi-line strings.
+    """
+    if (auto_set, enter_set) not in MultilineTextEditors:
+        from traitsui.api import TextEditor
+
+        MultilineTextEditors[auto_set, enter_set] = TextEditor(
+            multi_line=True, auto_set=auto_set, enter_set=enter_set
+        )
+
+    return MultilineTextEditors[auto_set, enter_set]
+
+
+def bytes_editor(auto_set=True, enter_set=False, encoding=None):
+    """ Factory function that returns a text editor for bytes.
+    """
+    if (auto_set, enter_set, encoding) not in BytesEditors:
+        from traitsui.api import TextEditor
+
+        if encoding is None:
+            format = bytes.hex
+            evaluate = bytes.fromhex
+        else:
+            format = partial(bytes.decode, encoding=encoding)
+            evaluate = partial(str.encode, encoding=encoding)
+
+        BytesEditors[(auto_set, enter_set, encoding)] = TextEditor(
+            multi_line=True,
```

Comment:
Probably unrelated to this PR (especially since it's just moving this code to a new place), but what's the effect of `multi_line=True` here? In manual testing, I just get a single-line editor.

Asking because I'm wondering how newlines get interpreted in hex mode.

Reply:
And the answer is that `bytes.fromhex` ignores spaces between bytes in Python 3.6, and allows general whitespace between bytes in Python >= 3.7.